### PR TITLE
deferred actions: load all values from deferred resources when building eval context

### PR DESCRIPTION
**Pull request by liamcervante**
This PR fixes a bug in deferred actions when they are referenced but haven't been added to state. Reasons why a deferred action might not be in state are:

- It's a data source
- It's creating a resource for the first time

Essentially, the evaluate functionality for resources looks at the state to work out the base set of values. These values were then overridden by values from the deferrals or from the plan when appropriate. But, this means that if a deferred resource wasn't already in state (eg. it's just being created) then no values would be added to the context for it. Now, we precompute the set of available instances by looking at the deferrals and state in isolation. We add all the values that are being deferred first, and then add values from the state/plan only if they haven't already been added. This ensures that even if there is nothing in the state for a deferred resource then we'll still capture values for it in the context.

This means that deferred actions are actually propagating their data properly for the first time, and this exposed a missing feature in outputs that reference deferred actions. These are supposed to be returned as null, but this was only happening because we weren't properly referencing deferred resources and the outputs were naturally returning unknown as a result. Now, the output node consults the deferred actions to check if any of the nodes it references were deferred, and replaces the output value with null properly.

To facilitate this, I changed the interface for `GraphNodeAttachDependencies` so that it can also be implemented by the output nodes. The functions for the plan and apply nodes must now manually remove themselves, but the same functionality can be shared for working out if an output node or a plan/apply node should be deferred because of dependencies.

### CODE REVIEW

File: internal/terraform/node_output.go [hunk 1]
```diff
@@ -781,10 +795,10 @@ func (n *NodeApplyableOutput) setValue(namedVals *namedvals.State, state *states
...
 		if n.Addr.Module.IsRoot() {
 			val, _ = val.UnmarkDeep()
-			if val.IsKnown() && !val.IsWhollyKnown() {
-				// If the value is partially unknown, act like it's fully
-				// unknown and store a null value in state. This can happen if
-				// an output references a deferred value.
+			if deferred.DependenciesDeferred(n.Dependencies) {
+				// If the output is from deferred resources then we return a
+				// simple null value representing that the value is really
+				// unknown as the dependencies were not properly computed.
```
Comment (nfagerlund) [inline internal/terraform/node_output.go:801]: Ah -- nice. I think this is more coherent.

**Review by nfagerlund — APPROVED** [review]
I'll defer to James on the performance and interface comments, but in terms of the main logic and the deferral lifecycle, this looks correct to me. Thank you for catching this!

File: internal/terraform/transform_reference.go [hunk 2]
```diff
@@ -52,8 +52,8 @@ type GraphNodeImportReferencer interface {
 }
 
 type GraphNodeAttachDependencies interface {
-	GraphNodeConfigResource
 	AttachDependencies([]addrs.ConfigResource)
+	Name() string
```
[the remark this answers is not in the document]
Comment (liamcervante) [reply internal/terraform/transform_reference.go]: Done!

File: internal/terraform/node_resource_plan.go [hunk 3]
```diff
@@ -73,7 +73,14 @@ func (n *nodeExpandPlannableResource) expandsInstances() {
 
 // GraphNodeAttachDependencies
 func (n *nodeExpandPlannableResource) AttachDependencies(deps []addrs.ConfigResource) {
-	n.dependencies = deps
+	n.dependencies = make([]addrs.ConfigResource, 0)
```
[the remark this answers is not in the document]
Comment (liamcervante) [reply internal/terraform/node_resource_plan.go]: I ended up pushing the check back into the AttachDependencies via checking for if the source is a `GraphNodeConfigResource` directly. This avoids the whole problem, so seemed like the best thing to do.

**Pull request merged**